keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,374 @@
+
+## Related issues
+
+- [Hardware topology awareness at node level (including NUMA)][numa-issue]
+- [Support Container Isolation of Hugepages][hugepage-issue]
+- [Support isolating memory to single NUMA node][memory-issue]
+
+## Goals
+
+- To guarantee isolation of memory and hugepages to single NUMA node for containers in the Guaranteed Pod.
+
+- To provide memory and hugepages topology hint to the Topology Manager. It will allow Topology Manager to allocate memory, hugepages, CPUs, and PCI devices on the same NUMA node.
+
+## Non-Goals
+
+- Supporting cross NUMA affinity is out of scope.
+
+- Isolating memory & hugepage to a NUMA node for multiple containers is out of scope.
+
+- Updating scheduler, pod spec is out of scope at this point.
+
+- This proposal only focus on Linux based systems.
+
+## User Stories
+
+### Story 1 : Networking Acceleration using DPDK
+
+- The system such as real-time trading system and 5G system, which requires networking acceleration, uses DPDK to guarantee low latency of packet processing. DPDK(Data Plane Development Kit) is set of libraries to accelerate packet processing on userspace. DPDK requires dedicated resources(such as exclusive cpu usage, hugepages and DPDK compatile Network Interface Card) and alignment of resources to sigle NUMA node to accelerate its performance. For this reason, there should be a way to guarantee a resource reservation of memory and hugepage as well as other computing resources (i.e. cpu core , NIC) from a single NUMA node for DPDK-based containers.
+
+### Story 2 : Database
+
+- Databases(Oracle, PostgreSQL, MySQL, MongoDB, SAP) consumes massive Memory and Hugepages, in order to reduce memory access latency and improve performance dramatically resources(CPU, Memory, Hugepages, I/O Devices) should be aligned.
+
+# Proposal
+
+## Proposed Changes
+
+### New Component: Memory Manager
+
+Memory Manager is a new component of Kubelet and a new hint provider of Topology Manager for memory and hugepage. It provides NUMA node affinity to the Topology Manager to guarantee NUMA aware memory consumption for a container in the Guaranteed QoS Pod.
+
+When Guaranteed QoS Pod admission is requested to kubelet, Topology Manager asks NUMA node affinity of memory and hugepage to Memory Manager. Then Memory Manager calculates NUMA node affinity based on internal database(`Memory Map`) and sends the affinity back to the Topology Manager to help the Topology Manager to figure out appropriate NUMA node where adequate resources are available for a given container.
+
+Note:
+- Memory Manager only support single NUMA node affinity, because Linux kernel doesn`t have such mechanism to reserve memory or set memory limit per NUMA node.
+
+Then, while the container is being deployed, Memory Manager reserves requested amount of memory on the internal database(`Memory Map`). Then it sets [CPUSET enforcement](#cpuset-enforcement) to consume memory from designated NUMA node, that Topology Manager selected.
```

Comment:
Can you please specify when should the manager pin `cpuset.mem` to te specific NUMA node, I think it makes sense to pin only in case when you already have CPU pinning, it means:
- CPU manager policy set to `static`
- container CPU and request limit equals and has an integer value

Reply:
Certainly I'd expect the most important high-performance use-case to be where we have CPU, PCI device, and memory/hugepages all coming from the same host NUMA node.

Reply:
I agree, and it will be done via topology hints, I just unsure when we should pin pod memory to the specific NUMA node, we should not do it for every pod.

Reply:
`To guarantee isolation of memory and hugepages to single NUMA node for containers in the Guaranteed Pod.`

Basically, Memory Manager will pin a memory when a Pod has guaranteed QoS, it means MM wil lpina memory for all containers in a Guaranteed QoS Pod.

Reply:
Hm, but if do not pin the CPU what the point to pin the memory?

Reply:
My additional plan is :
- adding new CPU Policy which supports NUMA aware cpu sharing. 
- adding new Topology Policy which support pod level single numa alignment.
(I`m not sure I will add pod-level-single-numa-policy.go literally or add new Policy using CRD to dsicribe the requirement of Pod resourcd topology and the level of topology such as {single-numa, container}, {single-numa, pod})

Reply:
`Hm, but if do not pin the CPU what the point to pin the memory?`
=> Once Memory Manager provides a hint for memory, topolgoy manager will determine the NUMA node(point) under configured policy.

Reply:
What about the case, when you have a container which requests a shared CPUs in the pod which has guaranteed QoS and does not have requests for other PCI devices, will you pin it memory anyway? It can always choose NUMA node 0 for example and it will bring NUMA node 0 to be exhausted.

Reply:
hm, your example also make sence when we use static policy of cpu manager.
But adding a dedicate logic per cpu policy on Memory Manager also doesnt make sence.

My goal in this year is very simple.
Making all containers in a pod(has Guaranteed QoS)  comsumes compute resource from a single NUMA node.
compute resources includes: exlusive cpu, shared cpu, memory, hugepage, nic, gpu, fpga, etc

Reply:
If you care about performance then all resources should be NUMA-aligned.  If you don't care about performance then it doesn't matter.

This is why I think that NUMA-alignment should be an attribute of the Pod (or even the container) and not the node.

That said, in order to properly track per-NUMA-node resources you can't have anything that "floats" across the whole node consuming resources unexpectedly from both NUMA nodes.   The prime example of this is non-hugepage memory.  By default Linux will allocate memory on the NUMA node that the request was made from...if a container can run on CPUs from both NUMA nodes then it can potentially allocate memory from both NUMA nodes, which makes it impossible to track how much memory is available for Kubernetes on each NUMA node.

Reply:
> This is why I think that NUMA-alignment requirements should be an attribute of the Pod (or even the container) and not the node.

+1